ls1intum/Artemis
# [Bugfix] Fix issue with template and solution participation

### Checklist

### Motivation and Context


There is an issue with the template and solution repository. In case of a new TEST submission the created result gets attached to the wrong submission. This is why the client shows that there is now result and shows a warning with a trigger new build button.

The issue is shown in the following picture: 
![image](https://user-images.githubusercontent.com/33484318/107330040-556e9c00-6ab1-11eb-871a-16bb1450480c.png)


### Description

When a result is created it gets attached to the first submission which includes the referenced commit hash. This was now changed by using the latest submission (sorted by submission date).

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
4. Verify that both results are shown correctly with no `red trigger again` button. Check in Programming Exercise Details and Online Editor.
5. Now change the Grading of the tests (e.g. set weight from any testcase from 1 to 2) and save. This will trigger the solution and template repository to build. (See other test steps below)
6. Go back to Programming Exercise Details and Online Editor. After the build of solution and template have finished verify that the change was correctly applied and no `red trigger again` button appears.

Further tests, which should be done in step 5):
- Reset the Grading by clicking the `reset` button.
- Apply changes to the SCA Categories (only if SCA is enabled)
- Open the test repository with `Edit in Editor` and push a change by e.g. removing a testcase in `SortingExampleBehaviorTest.java`

**PLEASE ALSO TEST IT LOCALLY ON YOUR MACHINE WITH THE LOCAL SETUP AND CHECK THE DATABASE**

Note: After all steps check the database tables `submission` and `result`.  
Only changes to the test repository will create a new `TEST` submission with a result for the solution and template participation.  
The changes to the grading and SCA categories will not create a new submission! They only trigger a build and therefore create a new result for the latest solution and template submission. See also screenshots below.

### Test Coverage



No changes.

### Screenshots



## Commit changes to test repository:
### Triggers solution repo to build:
![image](https://user-images.githubusercontent.com/33484318/107331184-cd899180-6ab2-11eb-980a-aec528428381.png)
![image](https://user-images.githubusercontent.com/33484318/107331813-a7182600-6ab3-11eb-8d25-6d29f2db001f.png)

### Afterwards the template repo build gets triggered:
![image](https://user-images.githubusercontent.com/33484318/107331324-04f83e00-6ab3-11eb-86a3-1ac46e6fe01b.png)

### Correct result in Programming Details:
![image](https://user-images.githubusercontent.com/33484318/107331039-a0d57a00-6ab2-11eb-971b-c9b39a27a60b.png)

## Database changes:
### Changes to test repo 
#### The exercise creation will create two MANUAL submissions and every push into the test repository will create two TEST submissions. In the screenshot below we created the exercise and did one push into the test repository (removed a test):
![image](https://user-images.githubusercontent.com/41108487/107499104-d783c100-6b94-11eb-9d6c-78bb470641eb.png)

#### For all modifications (changing grading, or SCA) only a new result should be generated which is connected to the lastest submission of the participation:
![image](https://user-images.githubusercontent.com/41108487/107499208-f5e9bc80-6b94-11eb-9c95-9c4d74f3b3a6.png)
